pyopensprinkler/__init__.py

```diff
@@ -1,84 +1,165 @@
 """Main OpenSprinkler module."""
 
+import hashlib
 import json
 import urllib
 
 import httplib2
 from backoff import expo, on_exception
 
-from pyopensprinkler.device import Device
 from pyopensprinkler.program import Program
 from pyopensprinkler.station import Station
 
-_HTTP = httplib2.Http()
 
+class Controller(object):
+    """OpenSprinkler Controller"""
 
-class OpenSprinkler(object):
-    """Represent the OpenSprinkler API."""
+    def __init__(self, url, password, client_opts=None):
+        """OpenSprinkler Controller initializer."""
 
-    def __init__(self, host, md5password):
-        """OpenSprinkler class initializer."""
-        self._host = host
-        self._md5password = md5password
-        self._baseUrl = f"http://{self._host}"
-        self._programs = []
-        self._stations = []
+        if client_opts is None:
+            client_opts = {}
+
+        self._password = password
+        self._md5password = hashlib.md5(password.encode('utf-8')).hexdigest()
+        self._baseUrl = url.strip("/")
+        self._programs = {}
+        self._stations = {}
         self._state = None
 
-        self.update()
-        self.device = Device(self)
+        client = httplib2.Http()
+        client.follow_all_redirects = True
+
+        if client_opts.http_username is not None:
+            client.add_credentials(client_opts.http_username, client_opts.http_password)
+
+        if client_opts.verify_ssl is not None:
+            client.disable_ssl_certificate_validation = not client_opts.verify_ssl
+
+        self._http_client = client
 
     def request(self, path, params=None):
         if params is None:
             params = {}
-        """Make a request from the API."""
+        """Make a request to the API."""
         params["pw"] = self._md5password
         qs = urllib.parse.urlencode(params)
+        url = '{}{}?{}'.format(self._baseUrl, path, qs)
```

Comment:
seems like the `f"{}"` style is preferred

Reply:
I'll update this. Another massive update coming shortly..